### PR TITLE
Add optional case_info nodes for HERON

### LINKED ISSUE #38: [TASK] Allow users to specify "case-info" constants
--------
Issue Description
--------
Users should be able to provide information pertaining to a case that would be valuable to have at the end of the RAVEN stack. This could be information that would help distinguish between runs when the data contains a multilevel structure. 

For example, If a user is examining effect on marginal costs across a set of varying market demand scenarios for a set of different markets, a user should be able to specify something like:

```xml
<Case name="Sweep_Runs">
...
  <caseinfo>
    <state>California</state>
    <scenario>CarbonTax</scenario>
  </caseinfo>
...
</Case>
```

This would allow HERON's template builder to handle a broader range of cases.

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or task?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [x] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by dylanjm**
--------
Pull Request Description
--------
##### What issue does this change request address?

Closes #38

##### What are the significant changes in functionality due to this change request?

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/HERON/wiki/Code-Standards) for details.
- [x] 4. Automated Tests should pass.
- [x] 5. If significant functionality is added, there must be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large tes.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added, the the analytic documentation must be updated/added.
- [x] 9. If any test used as a basis for documentation examples have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

**Comment by dylanjm** [comment]
@PaulTalbot-INL I'm able to run a few modified test-cases containing `<label>` nodes. But I'm not sure where we want it to end up. I thought it may end up in `disp_results.csv` but it doesn't.

File: src/Cases.py [hunk 1]
```diff
@@ -46,6 +46,20 @@ def get_input_specs(cls):
     input_specs.addParam('name', param_type=InputTypes.StringType, required=True,
         descr=r"""the name by which this analysis should be referred within HERON.""")
 
+    # Optional Identifier Nodes
+    label_specs = InputData.parameterInputFactory(
+        name='label',
+        ordered=False,
+        descr=r"""provides static label information to the model; unused in computation."""
```
Comment (PaulTalbot-INL) [inline src/Cases.py]: Can we specify here how these are accessed in the Functions? This is all the user manual they're going to get on this subject.

File: src/Cases.py [hunk 2]
```diff
@@ -460,7 +484,7 @@ def _modify_outer(self, template, components, sources):
     ###################
     run_info = template.find('RunInfo')
     case_name = self.get_working_dir('outer') #self.string_templates['jobname'].format(case=self.name, io='o')
-    ooooooooo # I don't think this gets run!
+    # ooooooooo # I don't think this gets run!
```
Comment (PaulTalbot-INL) [inline src/Cases.py]: does it get run?

File: templates/template_driver.py [hunk 3]
```diff
@@ -252,6 +252,13 @@ def _modify_outer_vargroups(self, template, components):
     # capacities
     caps = var_groups[0]
     caps.text = ', '.join('{}_capacity'.format(x.name) for x in components)
+    if case.get_labels():
+      case_labels = ET.SubElement(var_groups, 'Group', attrib={'name': 'GRO_case_labels'})
+      case_labels.text = ', '.join([f'{key}_label' for key in case.get_labels().keys()])
+      for node in template.find('DataObjects'):
+        if node.get('name') == 'grid':
+          input_node = node.find('Input')
+          input_node.text += ', GRO_case_labels'
```
Comment (PaulTalbot-INL) [inline templates/template_driver.py:261]: I don't think we need these in the outer/inner now, right? If they're on "case" they should be accessible already in the `meta` in the Functions, I believe.

**Review by PaulTalbot-INL — COMMENTED** [review]
A couple comments below.

I think the ultimate goal is to have these available in a Function method via `case.get_labels()` so we'll want to add a test that checks they can be obtained there.

File: src/Cases.py (hunk 2 above)
Comment (dylanjm) [reply src/Cases.py]: Nope, but was showing as an error in my IDE hah, shall I remove this?

File: templates/template_driver.py (hunk 3 above)
Comment (dylanjm) [reply templates/template_driver.py:261]: Do we want the case_info in the sweep output file? This was my initial way of getting it to show up there. Would there be a more straight-forward way?

File: src/Cases.py (hunk 2 above)
Comment (PaulTalbot-INL) [reply src/Cases.py]: it's fine, I was curious for my own sake. I'd leave the comment as you have it, though, so the next guy to come along can be aware.

EDIT: I just saw you already removed it; that's fine, although we need to address this whenever we start worrying about code coverage. I've been neglecting that so far.

File: templates/template_driver.py (hunk 3 above)
Comment (PaulTalbot-INL) [reply templates/template_driver.py:261]: Hm, interesting. Yeah, I see what you mean, good point. If this gets it in the output csv for the user automagically, that's good.

Comment (dylanjm) [reply templates/template_driver.py:261]: It's also showing up in the `disp_debug.csv` which I can take out if it wouldn't be necessary there.

File: src/Cases.py (hunk 2 above)
Comment (dylanjm) [reply src/Cases.py]: I was curious about this. It actually looks like `template_driver.py` duplicates all these functions. I noticed then when I first tried to add the aliases for the labels in these functions and they didn't show up. It also looks like they are nearly identical to the `template_driver.py` functions. Perhaps it's dead code we can remove?

Comment (PaulTalbot-INL) [reply src/Cases.py]: I think at one point Case was meant to also be the template writer, then I scrapped that for scope and bloat reasons. I suspect you're right.

**Comment by PaulTalbot-INL** [comment]
civet seems to be locked up right now; hopefully that's resolved soon.

**Comment by dylanjm** [comment]
@PaulTalbot-INL I added an integration test which is just an extended `production_flex` test. Let me know if that's sufficient.

**Comment by PaulTalbot-INL** [comment]
Yeah, it looks fine. It can become a great test in the future for testing what gets sent into the Functions, so standalone test is fine. I think we're just waiting on some documentation and the automatic tests, or did I miss the docs being added?

**Comment by dylanjm** [comment]
@PaulTalbot-INL I updated the documentation, but I just put all the points together on referencing `meta['HERON']['Case']`. I'm updating that now. Other than that, after the tests pass it should be good to go.

File: tests/integration_tests/labels/transfers.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  return data, meta
+
+def flex_price(data, meta):
+  sine = meta['HERON']['RAVEN_vars']['Signal']
+  t = meta['HERON']['time_index']
+  # DispatchManager
+  # scale electricity consumed to flex between -1 and 1
+  amount = - 2 * (sine[t] - 0.5)
+  data = {'reference_price': amount}
+  return data, meta
```
Comment (PaulTalbot-INL) [inline tests/integration_tests/labels/transfers.py:24]: Do you retrieve the labels anywhere as part of this test? I think that would show it's doing what it's supposed to.

You can cut out most of the comments in this file too, if you want.

**Review by PaulTalbot-INL — CHANGES_REQUESTED** [review]

File: tests/integration_tests/labels/transfers.py (hunk 4 above)
Comment (dylanjm) [reply tests/integration_tests/labels/transfers.py:24]: That should be updated now.

Comment (PaulTalbot-INL) [reply tests/integration_tests/labels/transfers.py:24]: Hm, interesting. In general I don't think we'd want to pass those back, but I think we'll happily ignore it even if we do. This mechanically works fine.

**Review by PaulTalbot-INL — APPROVED** [review]

**Pull request merged**